nimble/ll: Add Channel Sounding

mkasenberg opened the pull request.

mkasenberg committed:
nimble/ll: Add Channel Sounding state machine

Initialize the CS state machine with the connection state machine.

mkasenberg committed:
nimble/ll: Add Channel Sounding LL CTRL opcodes

Add lengths and opcodes of Channel Sounding LL CTRL PDUs.

mkasenberg committed:
nimble/ll: Add Channel Sounding capabilities

Implement the LE CS Read Local Supported Capabilities command.
Enable basic CS capabilities in local controller.

mkasenberg committed:
nimble/ll: Add CS Capabilities Exchange procedure

Implements:
- LE CS Read Remote Supported Capabilities command,
- LE CS Write Cached Remote Supported Capabilities command.

mkasenberg committed:
nimble/hci: Update Channel Sounding HCI

There were changes to HCI between the draft and final version of the
Bluetooth 6.0 specification.

mkasenberg committed:
nimble/ll: Add LE CS Set Default Settings command

The HCI_LE_CS_Set_Default_Settings command is used by a Host to set
default CS settings in the local Controller for the connection.

mkasenberg committed:
nimble/ll: Add CS Mode-0 FAE Table Request procedure

Implements:
- LE CS Read Remote FAE Table command
- LE CS Write Cached Remote FAE Table command

mkasenberg committed:
nimble/ll: Add CS Configuration procedure

Implements:
- LE CS Create Config command
- LE CS Remove Config command

mkasenberg committed:
nimble/ll: Add CS Security Start procedure

Implements LE CS Security Enable command.

mkasenberg committed:
nimble/ll: Add LE CS Set Procedure Parameters command

mkasenberg committed:
nimble/ll: Add LE CS Set Channel Classification command

mkasenberg committed:
nimble/ll: Add CS Start procedure

mkasenberg committed:
nimble/ll: Add initial Channel Sounding scheduling

sjanc commented on nimble/controller/src/ble_ll_cs.c:
```diff
@@ -26,75 +26,956 @@
+{
+    uint8_t config_id = connsm->cssm->config_req_id;
+    uint8_t action = connsm->cssm->config_req_action;
+    const struct ble_ll_cs_config *conf = &connsm->cssm->config[config_id];
```
is config_id already validated here?

sjanc commented on nimble/controller/src/ble_ll_cs.c:
```diff
@@ -26,75 +26,956 @@
+        return BLE_ERR_UNK_CONN_ID;
+    }
+
+    conf = &connsm->cssm->config[cmd->config_id];
```
config_id should be validated before being used for indexing

sjanc reviewed (commented):
look good, few nitpicks here and there

mkasenberg commented on nimble/controller/src/ble_ll_cs.c:
```diff
@@ -26,75 +26,956 @@
+int
+ble_ll_cs_rx_fae_req(struct ble_ll_conn_sm *connsm, uint8_t *rspbuf)
+{
+    memcpy(rspbuf, connsm->cssm->local_fae_table, 72);
```
Good point, there wasn't ;)

andrzej-kaczmarek commented on nimble/controller/src/ble_ll_cs.c:
```diff
@@ -26,75 +26,973 @@
+    struct ble_ll_cs_drbg_ctx *drbg_ctx = &connsm->cssm->drbg_ctx;
+
+    if (!IS_PENDING_CTRL_PROC(connsm, BLE_LL_CTRL_PROC_CS_SEC_START)) {
+        /* Should never happen */
```
in bluetooth it will happen ;)
just change the comment to "ignore" or smth

andrzej-kaczmarek commented on nimble/controller/src/ble_ll_cs.c:
```diff
@@ -26,75 +26,973 @@
+void
+ble_ll_cs_rx_capabilities_rsp(struct ble_ll_conn_sm *connsm, uint8_t *dptr)
+{
+    ble_ll_cs_update_rem_capabilities(connsm, dptr);
```
this should verify data in rsp and return error if needed

mkasenberg commented:
Just rebase

mkasenberg committed:
nimble/ll: Add unit tests for Channel Sounding CSA #3C

mkasenberg committed:
nimble/ll: Add BLE_LL_CS_LOG

The spec does not define a mechanism to inform which CS parameter the
controller rejects during CS start parameters negotiation, so a few
logs on the controller side will definitely be useful.

mkasenberg committed:
nimble/ll: Add unit tests for CS parameters validation

mkasenberg commented:
Contains UPF fixes now
